Allow attack orders to preempt move completion for turreted units.

pchote opened the pull request:
Fixes #14824.

This PR ~~(the last commit, after the changes from #16067)~~ adds an `OnQueueAttackActivity` notification that is called when an attack order is added to the activity queue.

`AttackFollow` uses this notification to set the target if the order is not queued.  This allows turrets to start tracking immediately, instead of waiting for an existing move order to complete and the attack order to tick for the first time.

This sits in the middle of the other targeting changes, so makes sense to include in the playtest. The improvement in perceived responsiveness should help soften the blow from turrets no longer tracking targets that are outside of range.

matjaeck commented:
On an idle unit, Move command -> Attack order -> Move command will make the unit's turret point at the target and moving the unit to the location of the last move command but it wont fire at the target in range. 

<details>
  <summary>GIF</summary>

![bildschirmaufnahme2019012600](https://user-images.githubusercontent.com/37185497/51778036-1d7da180-2100-11e9-8ed1-5f313c5cecd0.gif)

</details>

pchote commented:
Fixed.

matjaeck commented:
Units sometimes won't attack structures when targeted under fog. This is easily reproducible for defence structures and happens for other structures, too, but I don't know how to reproduce exactly. Something is causing inconsistent behavior.

<details>
  <summary>GIF not attacking defences</summary>

![1](https://user-images.githubusercontent.com/37185497/51781504-ad7c1500-2119-11e9-815d-14f1fcae1f0a.gif)

</details>

<details>
  <summary>GIF some tanks attack, some do not</summary>

![5](https://user-images.githubusercontent.com/37185497/51781846-026e5a00-211f-11e9-9b56-8b3d3110a73f.gif)




</details>

matjaeck commented:
I can still reproduce this on #16112 for buildings: build a medium tank with client one and send it out of your base. Build a power plant of client 2 next to where the medium tank is. Move the medium tank away from the power plant so that it becomes hidden under fog. Issue an attack order and following move command on it - the medium tank will pass the power plant without attacking.

pchote commented:
> GIF not attacking defences

Are you sure that you testing the latest version of the PR? I can't reproduce this.

matjaeck commented:
> Are you sure that you testing the latest version of the PR? I can't reproduce this.

> This looks the same as the first issue you raised - works for tanks that are ordered to attack a visible structure, doesn't work for tanks that are ordered to attack a frozen one. This should already be fixed!

As documented in the following GIF, there still are issues with targeting structures under fog where the unit won't attack when passing by. In the prior GIFs I intentionally captured the engine version but to make absolutely sure and document that I'm testing the right thing (could be that the version string is a leftover and not accurate) I now captured fetching the PR, which made the GIF a lot bigger. In this GIF it looks like that the unit fails to target after it has killed a target but I'm not sure if that is the repro case. 

https://imgur.com/a/mxC3Gy7

teinarss commented:
> 
> 
> > GIF not attacking defences
> 
> Are you sure that you testing the latest version of the PR? I can't reproduce this.
> 
> > GIF some tanks attack, some do not
> 
> This looks the same as the first issue you raised - works for tanks that are ordered to attack a visible structure, doesn't work for tanks that are ordered to attack a frozen one. This should already be fixed!

I can reproduce this also.

pchote commented:
I finally discovered the repro case here: it only happens if you build the pillbox *after* building the tank.

This happens because of an inconsistency in the way `FrozenUnderFog` handles visibility updates, which means you can catch actors in the middle of a tick where the game considers neither the FrozenActor nor the real actor to be visible.  This must be a really ancient bug that was then made worse by #15866.

I'm not yet sure whether it is best to try and include a fix here, in #16067, or in its own PR.

pchote committed:
Fix FrozenUnderFog / FrozenActor visibility consistency.

This fixes cases where both objects return visible / not
when queried at the wrong time during a tick.

pchote commented:
I added the fix here, even though this isn't related to this PR, because this provides the simplest repro case for testing.

teinarss commented:
Works now.

matjaeck commented:
Works for me too and looks ingame-wise good to me now :+1:

pchote commented:
Rebased now that the dependency has been merged.  The code changes should be significantly more readable now.

reaperrr commented on OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs:
```diff
@@ -81,6 +90,19 @@ public override Activity GetAttackActivity(Actor self, Target newTarget, bool al
+		public override void OnQueueAttackActivity(Actor self, Target target, bool queued, bool allowMove, bool forceAttack)
+		{
+			// If not queued we know that the attack activity will run next
+			// we can improve responsiveness for turreted actors by preempting
```
Readability nit: It looks like this starts a new sentence (or at least should), so the 'We' should use a capital W.

reaperrr reviewed (dismissed):
Code looks good to me, LGTM once that one comment is adressed

pchote committed:
Allow attack orders to preempt move completion for turreted units.

pchote commented:
Fixed.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog/_compare/b172a446a004d0bb049dcc1767c9758615ae2369...4ccb0a74928db34ac7a3483fdefd253bf54a945d

and picked to prep.
